<a href="https://colab.research.google.com/github/BxMild/bsc_dpdm24/blob/main/Crime_case.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [22]:
import pandas as pd
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [23]:
crime_data = pd.read_excel('/content/drive/MyDrive/Data_BSC_DPDM24/Crime-Dataset.xlsx')
crime_data

ภาค        จังหวัด                         ประเภทความผิด  \
0              NaN            NaN                                   NaN   
1    กรุงเทพมหานคร  กรุงเทพมหานคร  ความผิดเกี่ยวกับชีวิต ร่างกาย และเพศ   
2    กรุงเทพมหานคร  กรุงเทพมหานคร                            ฆ่าผู้อื่น   
3    กรุงเทพมหานคร  กรุงเทพมหานคร            ทำร้ายผู้อื่นถึงแก่ความตาย   
4    กรุงเทพมหานคร  กรุงเทพมหานคร                             พยายามฆ่า   
..             ...            ...                                   ...   
535         ภาคใต้       นราธิวาส            ทำร้ายผู้อื่นถึงแก่ความตาย   
536         ภาคใต้       นราธิวาส                             พยายามฆ่า   
537         ภาคใต้       นราธิวาส                         ทำร้ายร่างกาย   
538         ภาคใต้       นราธิวาส                      ข่มขืนกระทำชำเรา   
539         ภาคใต้       นราธิวาส                                อื่น ๆ   

     รับแจ้ง_2560  จับกุม_2560  ผู้ต้องหา_2560  รับแจ้ง_2561  จับกุม_2561  \
0             NaN          NaN             NaN           NaN          NaN   
1          2727.0       2036.0          2503.0        2457.0       2345.0   
2            88.0         77.0           106.0          88.0        120.0   
3            48.0         39.0            57.0          32.0         38.0   
4           110.0         82.0           125.0          84.0         85.0   
..            ...          ...             ...           ...          ...   
535           1.0          0.0             0.0           1.0          4.0   
536          89.0         16.0            32.0          68.0         15.0   
537          56.0         26.0            31.0          31.0         17.0   
538          10.0          5.0            15.0           8.0          4.0   
539          11.0          7.0            10.0           5.0          3.0   

     ผู้ต้องหา_2561  รับแจ้ง_2562  ...  ผู้ต้องหา_2562  รับแจ้ง_2563  \
0               NaN           NaN  ...             NaN           NaN   
1            2484.0        2395.0  ...          2548.0        2119.0   
2             125.0          81.0  ...           131.0          63.0   
3              38.0          37.0  ...            43.0          34.0   
4              88.0          85.0  ...           102.0          72.0   
..              ...           ...  ...             ...           ...   
535             4.0           2.0  ...             1.0           0.0   
536            15.0          60.0  ...            30.0          43.0   
537            17.0          28.0  ...            17.0          20.0   
538             4.0          15.0  ...            11.0          15.0   
539             3.0           8.0  ...             5.0          31.0   

     จับกุม_2563  ผู้ต้องหา_2563  รับแจ้ง_2564  จับกุม_2564  ผู้ต้องหา_2564  \
0            NaN             NaN           NaN          NaN             NaN   
1         1897.0          2314.0        1993.0       1846.0          2459.0   
2           58.0            76.0          36.0         36.0            66.0   
3           31.0            39.0          28.0         27.0            52.0   
4           63.0           104.0          81.0         70.0           121.0   
..           ...             ...           ...          ...             ...   
535          0.0             0.0           1.0          1.0             3.0   
536         22.0            32.0          80.0         37.0            56.0   
537         20.0            24.0          13.0         13.0            19.0   
538         12.0            12.0           9.0          9.0            12.0   
539         27.0            32.0          48.0         42.0            64.0   

     รับแจ้ง_2565  จับกุม_2565  ผู้ต้องหา_2565  
0             NaN          NaN             NaN  
1          2363.0       2227.0          2947.0  
2            83.0         78.0           135.0  
3            38.0         36.0            47.0  
4           126.0        121.0           220.0  
..            ...          ...             ...  
535           4.0          3.0  

In [24]:
crime_data = crime_data.dropna()
crime_data

ภาค        จังหวัด                         ประเภทความผิด  \
1    กรุงเทพมหานคร  กรุงเทพมหานคร  ความผิดเกี่ยวกับชีวิต ร่างกาย และเพศ   
2    กรุงเทพมหานคร  กรุงเทพมหานคร                            ฆ่าผู้อื่น   
3    กรุงเทพมหานคร  กรุงเทพมหานคร            ทำร้ายผู้อื่นถึงแก่ความตาย   
4    กรุงเทพมหานคร  กรุงเทพมหานคร                             พยายามฆ่า   
5    กรุงเทพมหานคร  กรุงเทพมหานคร                         ทำร้ายร่างกาย   
..             ...            ...                                   ...   
535         ภาคใต้       นราธิวาส            ทำร้ายผู้อื่นถึงแก่ความตาย   
536         ภาคใต้       นราธิวาส                             พยายามฆ่า   
537         ภาคใต้       นราธิวาส                         ทำร้ายร่างกาย   
538         ภาคใต้       นราธิวาส                      ข่มขืนกระทำชำเรา   
539         ภาคใต้       นราธิวาส                                อื่น ๆ   

     รับแจ้ง_2560  จับกุม_2560  ผู้ต้องหา_2560  รับแจ้ง_2561  จับกุม_2561  \
1          2727.0       2036.0          2503.0        2457.0       2345.0   
2            88.0         77.0           106.0          88.0        120.0   
3            48.0         39.0            57.0          32.0         38.0   
4           110.0         82.0           125.0          84.0         85.0   
5          2102.0       1591.0          1955.0        1942.0       1854.0   
..            ...          ...             ...           ...          ...   
535           1.0          0.0             0.0           1.0          4.0   
536          89.0         16.0            32.0          68.0         15.0   
537          56.0         26.0            31.0          31.0         17.0   
538          10.0          5.0            15.0           8.0          4.0   
539          11.0          7.0            10.0           5.0          3.0   

     ผู้ต้องหา_2561  รับแจ้ง_2562  ...  ผู้ต้องหา_2562  รับแจ้ง_2563  \
1            2484.0        2395.0  ...          2548.0        2119.0   
2             125.0          81.0  ...           131.0          63.0   
3              38.0          37.0  ...            43.0          34.0   
4              88.0          85.0  ...           102.0          72.0   
5            1973.0        1875.0  ...          2008.0        1172.0   
..              ...           ...  ...             ...           ...   
535             4.0           2.0  ...             1.0           0.0   
536            15.0          60.0  ...            30.0          43.0   
537            17.0          28.0  ...            17.0          20.0   
538             4.0          15.0  ...            11.0          15.0   
539             3.0           8.0  ...             5.0          31.0   

     จับกุม_2563  ผู้ต้องหา_2563  รับแจ้ง_2564  จับกุม_2564  ผู้ต้องหา_2564  \
1         1897.0          2314.0        1993.0       1846.0          2459.0   
2           58.0            76.0          36.0         36.0            66.0   
3           31.0            39.0          28.0         27.0            52.0   
4           63.0           104.0          81.0         70.0           121.0   
5         1089.0          1297.0         790.0        764.0           972.0   
..           ...             ...           ...          ...             ...   
535          0.0             0.0           1.0          1.0             3.0   
536         22.0            32.0          80.0         37.0            56.0   
537         20.0            24.0          13.0         13.0            19.0   
538         12.0            12.0           9.0          9.0            12.0   
539         27.0            32.0          48.0         42.0            64.0   

     รับแจ้ง_2565  จับกุม_2565  ผู้ต้องหา_2565  
1          2363.0       2227.0          2947.0  
2            83.0         78.0           135.0  
3            38.0         36.0            47.0  
4           126.0        121.0           220.0  
5          1759.0       1669.0          2207.0  
..            ...          ...             ...  
535           4.0          3.0  

In [25]:
# prompt: ลบแถวที่ประเภทความผิดเป็น ความผิดเกี่ยวกับชีวิต ร่างกาย และเพศ

# Assuming 'ประเภทความผิด' is the column name containing crime type
crime_data = crime_data[crime_data['ประเภทความผิด'] != 'ความผิดเกี่ยวกับชีวิต ร่างกาย และเพศ']
crime_data


ภาค        จังหวัด               ประเภทความผิด  รับแจ้ง_2560  \
2    กรุงเทพมหานคร  กรุงเทพมหานคร                  ฆ่าผู้อื่น          88.0   
3    กรุงเทพมหานคร  กรุงเทพมหานคร  ทำร้ายผู้อื่นถึงแก่ความตาย          48.0   
4    กรุงเทพมหานคร  กรุงเทพมหานคร                   พยายามฆ่า         110.0   
5    กรุงเทพมหานคร  กรุงเทพมหานคร               ทำร้ายร่างกาย        2102.0   
6    กรุงเทพมหานคร  กรุงเทพมหานคร            ข่มขืนกระทำชำเรา         226.0   
..             ...            ...                         ...           ...   
535         ภาคใต้       นราธิวาส  ทำร้ายผู้อื่นถึงแก่ความตาย           1.0   
536         ภาคใต้       นราธิวาส                   พยายามฆ่า          89.0   
537         ภาคใต้       นราธิวาส               ทำร้ายร่างกาย          56.0   
538         ภาคใต้       นราธิวาส            ข่มขืนกระทำชำเรา          10.0   
539         ภาคใต้       นราธิวาส                      อื่น ๆ          11.0   

     จับกุม_2560  ผู้ต้องหา_2560  รับแจ้ง_2561  จับกุม_2561  ผู้ต้องหา_2561  \
2           77.0           106.0          88.0        120.0           125.0   
3           39.0            57.0          32.0         38.0            38.0   
4           82.0           125.0          84.0         85.0            88.0   
5         1591.0          1955.0        1942.0       1854.0          1973.0   
6          140.0           148.0         175.0        132.0           141.0   
..           ...             ...           ...          ...             ...   
535          0.0             0.0           1.0          4.0             4.0   
536         16.0            32.0          68.0         15.0            15.0   
537         26.0            31.0          31.0         17.0            17.0   
538          5.0            15.0           8.0          4.0             4.0   
539          7.0            10.0           5.0          3.0             3.0   

     รับแจ้ง_2562  ...  ผู้ต้องหา_2562  รับแจ้ง_2563  จับกุม_2563  \
2            81.0  ...           131.0          63.0         58.0   
3            37.0  ...            43.0          34.0         31.0   
4            85.0  ...           102.0          72.0         63.0   
5          1875.0  ...          2008.0        1172.0       1089.0   
6           133.0  ...           107.0         124.0        102.0   
..            ...  ...             ...           ...          ...   
535           2.0  ...             1.0           0.0          0.0   
536          60.0  ...            30.0          43.0         22.0   
537          28.0  ...            17.0          20.0         20.0   
538          15.0  ...            11.0          15.0         12.0   
539           8.0  ...             5.0          31.0         27.0   

     ผู้ต้องหา_2563  รับแจ้ง_2564  จับกุม_2564  ผู้ต้องหา_2564  รับแจ้ง_2565  \
2              76.0          36.0         36.0            66.0          83.0   
3              39.0          28.0         27.0            52.0          38.0   
4             104.0          81.0         70.0           121.0         126.0   
5            1297.0         790.0        764.0           972.0        1759.0   
6             111.0         127.0        105.0           124.0         151.0   
..              ...           ...          ...             ...           ...   
535             0.0           1.0          1.0             3.0           4.0   
536            32.0          80.0         37.0            56.0          72.0   
537            24.0          13.0         13.0            19.0          39.0   
538            12.0           9.0          9.0            12.0           9.0   
539            32.0          48.0         42.0            64.0          15.0   

     จับกุม_2565  ผู้ต้องหา_2565  
2           78.0           135.0  
3           36.0            47.0  
4          121.0           220.0  
5         1669.0          2207.0  
6          135.0           137.0  
..           ...             ...  
535          3.0             4.0  
536         39.0            60.0  
537         

In [27]:
crime_df = crime_data.melt(id_vars=['ภาค', 'จังหวัด', 'ประเภทความผิด'],
                    value_vars=['รับแจ้ง_2560', 'จับกุม_2560', 'ผู้ต้องหา_2560',
                                'รับแจ้ง_2561', 'จับกุม_2561', 'ผู้ต้องหา_2561',
                                'รับแจ้ง_2562', 'จับกุม_2562', 'ผู้ต้องหา_2562',
                                'รับแจ้ง_2563', 'จับกุม_2563', 'ผู้ต้องหา_2563',
                                'รับแจ้ง_2564', 'จับกุม_2564', 'ผู้ต้องหา_2564',
                                'รับแจ้ง_2565', 'จับกุม_2565', 'ผู้ต้องหา_2565'],
                    var_name='ปีและประเภท', value_name='จำนวน')

# แยกคอลัมน์ ปี และ ประเภท
crime_df[['ประเภท', 'ปี']] = crime_df['ปีและประเภท'].str.extract(r'([ก-๙]+)_(\d{4})')

# ลบคอลัมน์ปีและประเภทเดิม
crime_df = crime_df.drop(columns=['ปีและประเภท'])

# แสดงผลลัพธ์
crime_df.head()

ภาค        จังหวัด               ประเภทความผิด   จำนวน   ประเภท  \
0  กรุงเทพมหานคร  กรุงเทพมหานคร                  ฆ่าผู้อื่น    88.0  รับแจ้ง   
1  กรุงเทพมหานคร  กรุงเทพมหานคร  ทำร้ายผู้อื่นถึงแก่ความตาย    48.0  รับแจ้ง   
2  กรุงเทพมหานคร  กรุงเทพมหานคร                   พยายามฆ่า   110.0  รับแจ้ง   
3  กรุงเทพมหานคร  กรุงเทพมหานคร               ทำร้ายร่างกาย  2102.0  รับแจ้ง   
4  กรุงเทพมหานคร  กรุงเทพมหานคร            ข่มขืนกระทำชำเรา   226.0  รับแจ้ง   

     ปี  
0  2560  
1  2560  
2  2560  
3  2560  
4  2560

In [28]:
crime_df

ภาค        จังหวัด               ประเภทความผิด   จำนวน  \
0     กรุงเทพมหานคร  กรุงเทพมหานคร                  ฆ่าผู้อื่น    88.0   
1     กรุงเทพมหานคร  กรุงเทพมหานคร  ทำร้ายผู้อื่นถึงแก่ความตาย    48.0   
2     กรุงเทพมหานคร  กรุงเทพมหานคร                   พยายามฆ่า   110.0   
3     กรุงเทพมหานคร  กรุงเทพมหานคร               ทำร้ายร่างกาย  2102.0   
4     กรุงเทพมหานคร  กรุงเทพมหานคร            ข่มขืนกระทำชำเรา   226.0   
...             ...            ...                         ...     ...   
8311         ภาคใต้       นราธิวาส  ทำร้ายผู้อื่นถึงแก่ความตาย     4.0   
8312         ภาคใต้       นราธิวาส                   พยายามฆ่า    60.0   
8313         ภาคใต้       นราธิวาส               ทำร้ายร่างกาย    39.0   
8314         ภาคใต้       นราธิวาส            ข่มขืนกระทำชำเรา     7.0   
8315         ภาคใต้       นราธิวาส                      อื่น ๆ    13.0   

         ประเภท    ปี  
0       รับแจ้ง  2560  
1       รับแจ้ง  2560  
2       รับแจ้ง  2560  
3       รับแจ้ง  2560  
4       รับแจ้ง  2560  
...         ...   ...  
8311  ผู้ต้องหา  2565  
8312  ผู้ต้องหา  2565  
8313  ผู้ต้องหา  2565  
8314  ผู้ต้องหา  2565  
8315  ผู้ต้องหา  2565  

[8316 rows x 6 columns]

In [29]:
# prompt: export csv

crime_df.to_csv('crime_data_exported.csv', encoding='utf-8-sig', index=False)


In [30]:
crime_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8316 entries, 0 to 8315
Data columns (total 6 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   ภาค            8316 non-null   object 
 1   จังหวัด        8316 non-null   object 
 2   ประเภทความผิด  8316 non-null   object 
 3   จำนวน          8316 non-null   float64
 4   ประเภท         8316 non-null   object 
 5   ปี             8316 non-null   object 
dtypes: float64(1), object(5)
memory usage: 389.9+ KB


1. แนวคิดของ Association Rules
Association Rules หรือ กฎความสัมพันธ์ ใช้สำหรับวิเคราะห์ว่าประเภทอาชญากรรมใดมักเกิดร่วมกัน โดยเราจะใช้ Apriori Algorithm หรือ FP-Growth Algorithm เพื่อค้นหากฎ เช่น

"จังหวัดที่มีคดีฆาตกรรมสูง มักมีคดีพยายามฆ่าสูงเช่นกัน"
"จังหวัดที่มีคดีข่มขืนสูง มักมีคดีทำร้ายร่างกายสูง"

In [36]:
import pandas as pd

# โหลดข้อมูล
df = crime_df.copy()

# แปลงข้อมูลให้อยู่ในรูปแบบ Pivot Table
df_pivot = df.pivot_table(index=['จังหวัด', 'ปี'],
                           columns='ประเภทความผิด',
                           values='จำนวน',
                           aggfunc='sum').reset_index()

# แก้ไขชื่อคอลัมน์ให้อ่านง่ายขึ้น
df_pivot.columns.name = None  # ลบชื่อคอลัมน์ที่เกิดจาก pivot
df_pivot = df_pivot.rename(columns={'ฆ่าผู้อื่น': 'ฆาตกรรม', 'ทำร้ายผู้อื่น': 'ทำร้ายร่างกาย'})

# แสดงผลลัพธ์
df_pivot.head()


จังหวัด    ปี  ข่มขืนกระทำชำเรา  ฆาตกรรม  ทำร้ายผู้อื่นถึงแก่ความตาย  \
0  กระบี่  2560              31.0     96.0                         3.0   
1  กระบี่  2561              22.0    100.0                        18.0   
2  กระบี่  2562              36.0     51.0                        20.0   
3  กระบี่  2563              28.0     38.0                         6.0   
4  กระบี่  2564              34.0     29.0                        24.0   

   ทำร้ายร่างกาย  พยายามฆ่า  อื่น ๆ  
0          179.0      141.0    41.0  
1          255.0      131.0    41.0  
2          202.0       75.0    55.0  
3          158.0       91.0   121.0  
4           88.0       70.0   155.0

In [38]:
# prompt: แปลงข้อมูลในรูปแบบ one-hot encoding โดยเงื่อนไขคือ  1 คือ สูง 0 คือ ต้ำ

# Assuming crime counts are now spread across different crime type columns
# We'll categorize each crime type count into 'สูง' (1) and 'ต่ำ' (0) based on a threshold.

# Calculate the median across all crime type columns, excluding 'จังหวัด' and 'ปี'
crime_type_columns = [col for col in df_pivot.columns if col not in ['จังหวัด', 'ปี']]
threshold = df_pivot[crime_type_columns].stack().median()  # Using the median as the threshold

# Create new columns with one-hot encoding for each crime type
for crime_type in crime_type_columns:
    df_pivot[crime_type + '_onehot'] = (df_pivot[crime_type] >= threshold).astype(int)

# Now, for each crime type, there's a corresponding '_onehot' column
# with 1 for counts above or equal to the threshold and 0 otherwise.

df_pivot.head()


จังหวัด    ปี  ข่มขืนกระทำชำเรา  ฆาตกรรม  ทำร้ายผู้อื่นถึงแก่ความตาย  \
0  กระบี่  2560              31.0     96.0                         3.0   
1  กระบี่  2561              22.0    100.0                        18.0   
2  กระบี่  2562              36.0     51.0                        20.0   
3  กระบี่  2563              28.0     38.0                         6.0   
4  กระบี่  2564              34.0     29.0                        24.0   

   ทำร้ายร่างกาย  พยายามฆ่า  อื่น ๆ  ข่มขืนกระทำชำเรา_onehot  ฆาตกรรม_onehot  \
0          179.0      141.0    41.0                        0               1   
1          255.0      131.0    41.0                        0               1   
2          202.0       75.0    55.0                        0               1   
3          158.0       91.0   121.0                        0               0   
4           88.0       70.0   155.0                        0               0   

   ทำร้ายผู้อื่นถึงแก่ความตาย_onehot  ทำร้ายร่างกาย_onehot  พยายามฆ่า_onehot  \
0                                  0                     1                 1   
1                                  0                     1                 1   
2                                  0                     1                 1   
3                                  0                     1                 1   
4                                  0                     1                 1   

   อื่น ๆ_onehot  
0              0  
1              0  
2              1  
3              1  
4              1

In [39]:
# prompt: ทำ association rules

from mlxtend.frequent_patterns import apriori, association_rules

# Select only the one-hot encoded columns for association rule mining
onehot_cols = [col for col in df_pivot.columns if '_onehot' in col]
df_onehot = df_pivot[onehot_cols]

# Apply Apriori algorithm
frequent_itemsets = apriori(df_onehot, min_support=0.05, use_colnames=True)

# Generate association rules
rules = association_rules(frequent_itemsets, metric="lift", min_threshold=1)

# Display the rules
rules


/usr/local/lib/python3.11/dist-packages/mlxtend/frequent_patterns/fpcommon.py:161: DeprecationWarning: DataFrames with non-bool types result in worse computationalperformance and their support might be discontinued in the future.Please use a DataFrame with bool type
  warnings.warn(


antecedents  \
0    (ข่มขืนกระทำชำเรา_onehot)   
1             (ฆาตกรรม_onehot)   
2    (ข่มขืนกระทำชำเรา_onehot)   
3       (ทำร้ายร่างกาย_onehot)   
4    (ข่มขืนกระทำชำเรา_onehot)   
..                         ...   
173           (ฆาตกรรม_onehot)   
174     (ทำร้ายร่างกาย_onehot)   
175            (อื่น ๆ_onehot)   
176  (ข่มขืนกระทำชำเรา_onehot)   
177         (พยายามฆ่า_onehot)   

                                           consequents  antecedent support  \
0                                     (ฆาตกรรม_onehot)            0.534632   
1                            (ข่มขืนกระทำชำเรา_onehot)            0.361472   
2                               (ทำร้ายร่างกาย_onehot)            0.534632   
3                            (ข่มขืนกระทำชำเรา_onehot)            0.948052   
4                                   (พยายามฆ่า_onehot)            0.534632   
..                                                 ...                 ...   
173  (ข่มขืนกระทำชำเรา_onehot, ทำร้ายร่างกาย_onehot...            0.361472   
174  (ข่มขืนกระทำชำเรา_onehot, ฆาตกรรม_onehot, อื่น...            0.948052   
175  (ข่มขืนกระทำชำเรา_onehot, ฆาตกรรม_onehot, ทำร้...            0.610390   
176  (ฆาตกรรม_onehot, อื่น ๆ_onehot, ทำร้ายร่างกาย_...            0.534632   
177  (ข่มขืนกระทำชำเรา_onehot, ฆาตกรรม_onehot, อื่น...            0.549784   

     consequent support   support  confidence      lift  representativity  \
0              0.361472  0.279221    0.522267  1.444835               1.0   
1              0.534632  0.279221    0.772455  1.444835               1.0   
2              0.948052  0.532468    0.995951  1.050524               1.0   
3              0.534632  0.532468    0.561644  1.050524               1.0   
4              0.549784  0.409091    0.765182  1.391788               1.0   
..                  ...       ...         ...       ...               ...   
173            0.341991  0.225108    0.622754  1.820966               1.0   
174            0.225108  0.225108    0.237443  1.054795               1.0   
175            0.251082  0.225108    0.368794  1.468819               1.0   
176            0.246753  0.225108    0.421053  1.706371               1.0   
177            0.242424  0.225108    0.409449  1.688976               1.0   

     leverage  conviction  zhangs_metric   jaccard  certainty  kulczynski  
0    0.085966    1.336580       0.661583  0.452632   0.251822    0.647361  
1    0.085966    2.045170       0.482171  0.452632   0.511043    0.647361  
2    0.025609   12.831169       0.103347  0.560364   0.922065    0.778798  
3    0.025609    1.061621       0.925813  0.560364   0.058044    0.778798  
4    0.115159    1.917301       0.604897  0.605769   0.478434    0.754638  
..        ...         ...            ...       ...        ...         ...  
173  0.101488    1.744245       0.706063  0.470588   0.426686    0.640491  
174  0.011694    1.016175       1.000000  0.237443   0.015918    0.618721  
175  0.071850    1.186488       0.819231  0.353741   0.157176    0.632673  
176  0.093186    1.301063       0.889535  0.404669   0.231397    0.666667  
177  0.091827    1.282828       0.906065  0.396947   0.220472    0.669010  

[178 rows x 14 columns]

In [42]:
# prompt: แสดงสิบกฎที่มีความสัมพันธ์กันสูงสุด

# Display the top 10 rules with the highest lift
top_10_rules = rules.sort_values(by='lift', ascending=False).head(10)
top_10_rules

antecedents  \
153  (ฆาตกรรม_onehot, อื่น ๆ_onehot, ทำร้ายร่างกาย_...   
172        (ข่มขืนกระทำชำเรา_onehot, พยายามฆ่า_onehot)   
160  (ทำร้ายร่างกาย_onehot, อื่น ๆ_onehot, พยายามฆ่...   
165          (ข่มขืนกระทำชำเรา_onehot, ฆาตกรรม_onehot)   
164                    (ฆาตกรรม_onehot, อื่น ๆ_onehot)   
161  (ข่มขืนกระทำชำเรา_onehot, ทำร้ายร่างกาย_onehot...   
113                    (ฆาตกรรม_onehot, อื่น ๆ_onehot)   
112        (ข่มขืนกระทำชำเรา_onehot, พยายามฆ่า_onehot)   
110          (ข่มขืนกระทำชำเรา_onehot, ฆาตกรรม_onehot)   
171                  (อื่น ๆ_onehot, พยายามฆ่า_onehot)   

                                           consequents  antecedent support  \
153        (ข่มขืนกระทำชำเรา_onehot, พยายามฆ่า_onehot)            0.268398   
172  (ฆาตกรรม_onehot, อื่น ๆ_onehot, ทำร้ายร่างกาย_...            0.409091   
160          (ข่มขืนกระทำชำเรา_onehot, ฆาตกรรม_onehot)            0.400433   
165  (ทำร้ายร่างกาย_onehot, อื่น ๆ_onehot, พยายามฆ่...            0.279221   
164  (ข่มขืนกระทำชำเรา_onehot, ทำร้ายร่างกาย_onehot...            0.274892   
161                    (ฆาตกรรม_onehot, อื่น ๆ_onehot)            0.406926   
113        (ข่มขืนกระทำชำเรา_onehot, พยายามฆ่า_onehot)            0.274892   
112                    (ฆาตกรรม_onehot, อื่น ๆ_onehot)            0.409091   
110                  (อื่น ๆ_onehot, พยายามฆ่า_onehot)            0.279221   
171  (ข่มขืนกระทำชำเรา_onehot, ฆาตกรรม_onehot, ทำร้...            0.411255   

     consequent support   support  confidence      lift  representativity  \
153            0.409091  0.225108    0.838710  2.050179               1.0   
172            0.268398  0.225108    0.550265  2.050179               1.0   
160            0.279221  0.225108    0.562162  2.013325               1.0   
165            0.400433  0.225108    0.806202  2.013325               1.0   
164            0.406926  0.225108    0.818898  2.012397               1.0   
161            0.274892  0.225108    0.553191  2.012397               1.0   
113            0.409091  0.225108    0.818898  2.001750               1.0   
112            0.274892  0.225108    0.550265  2.001750               1.0   
110            0.411255  0.225108    0.806202  1.960343               1.0   
171            0.279221  0.225108    0.547368  1.960343               1.0   

     leverage  conviction  zhangs_metric   jaccard  certainty  kulczynski  
153  0.115309    3.663636       0.700159  0.497608   0.727047    0.694487  
172  0.115309    1.626738       0.866864  0.497608   0.385273    0.694487  
160  0.113299    1.646224       0.839454  0.495238   0.392549    0.684182  
165  0.113299    3.093766       0.698285  0.495238   0.676769    0.684182  
164  0.113248    3.274798       0.693800  0.492891   0.694638    0.686045  
161  0.113248    1.622861       0.848259  0.492891   0.383804    0.686045  
113  0.112652    3.262846       0.690155  0.490566   0.693519    0.684581  
112  0.112652    1.612299       0.846893  0.490566   0.379768    0.684581  
110  0.110277    3.037922       0.679660  0.483721   0.670828    0.676785  
171  0.110277    1.592419       0.832084  0.483721   0.372025    0.676785

🔎 สรุปผลจากการวิเคราะห์ความสัมพันธ์ของอาชญากรรม
จากตาราง Association Rule Mining (กฎความสัมพันธ์) ที่ได้จากข้อมูลอาชญากรรม สามารถสรุปผลสำคัญได้ดังนี้:

1. ความสัมพันธ์ระหว่าง "ข่มขืนกระทำชำเรา" กับ "ฆาตกรรม"
มีโอกาส 52.23% ที่หากเกิดเหตุ ข่มขืนกระทำชำเรา จะเกิด ฆาตกรรม ด้วย
ในทางกลับกัน มีโอกาส 77.25% ที่หากเกิด ฆาตกรรม จะมีเหตุ ข่มขืนกระทำชำเรา
ค่า Lift = 1.44 แสดงให้เห็นว่าโอกาสที่เหตุการณ์เหล่านี้จะเกิดขึ้นพร้อมกันสูงกว่าการเกิดขึ้นแบบสุ่ม
2. ความสัมพันธ์ระหว่าง "ข่มขืนกระทำชำเรา" กับ "ทำร้ายร่างกาย"
หากเกิด ข่มขืนกระทำชำเรา มีโอกาสสูงถึง 99.59% ที่จะเกิดเหตุ ทำร้ายร่างกาย ด้วย
อย่างไรก็ตาม แม้ว่าจะมีเหตุ ทำร้ายร่างกาย แต่มีเพียง 56.16% ที่จะเกี่ยวข้องกับ ข่มขืนกระทำชำเรา
ค่า Lift = 1.05 บ่งชี้ว่าความสัมพันธ์นี้มีอยู่จริงแต่ไม่ได้แข็งแกร่งมาก
3. ความสัมพันธ์ระหว่าง "ข่มขืนกระทำชำเรา" กับ "พยายามฆ่า"
หากเกิด ข่มขืนกระทำชำเรา มีโอกาส 76.52% ที่จะนำไปสู่ พยายามฆ่า
ค่า Lift = 1.39 บ่งชี้ว่าความสัมพันธ์นี้แข็งแกร่งกว่าค่าเฉลี่ย
📌 ข้อสังเกตและข้อสรุป
"ข่มขืนกระทำชำเรา" มักจะเกี่ยวข้องกับอาชญากรรมรุนแรงอื่นๆ เช่น ฆาตกรรม, ทำร้ายร่างกาย และพยายามฆ่า
"ทำร้ายร่างกาย" พบได้บ่อยและมักเกิดควบคู่กับคดีข่มขืน แต่ไม่ได้มีความสัมพันธ์ที่แข็งแกร่งในทางกลับกัน
"ข่มขืนกระทำชำเรา" กับ "พยายามฆ่า" มีความเชื่อมโยงที่สำคัญ ซึ่งแสดงให้เห็นว่าอาชญากรรมทั้งสองประเภทนี้มีแนวโน้มเกิดร่วมกันสูง
🎯 ข้อเสนอแนะ
เพิ่มมาตรการป้องกัน สำหรับอาชญากรรมประเภท ข่มขืนกระทำชำเรา เนื่องจากมีแนวโน้มจะนำไปสู่อาชญากรรมรุนแรงอื่น
เฝ้าระวังคดีทำร้ายร่างกาย เพราะแม้จะพบได้ทั่วไป แต่ไม่ได้มีความเชื่อมโยงที่แข็งแกร่งกับคดีข่มขืน อาจต้องใช้แนวทางแก้ปัญหาที่แตกต่าง
ตรวจสอบรายละเอียดในคดีข่มขืนและพยายามฆ่า อย่างใกล้ชิด เนื่องจากมีโอกาสสูงที่ทั้งสองเหตุการณ์จะเกิดขึ้นร่วมกัน